### PR TITLE
add forecast_opts()

### LINKED ISSUE #867: `forecast_opts()` or `add_horizon()`
Just surfacing a discussion mixed up with other discussions in https://github.com/epiforecasts/EpiNow2/issues/346#issuecomment-2265209320

In order to address #640 we need to decide how to specify accumulation for future data. Options on the table are (for an example of weekly forecasts):

```r
obs |>
  estimate_infections(forecast = forecast_opts(horizon = 14, accumulate = 7)
```

or

```r
obs |>
  add_horizon(horizon = 14, accumulate = 7) |>
  estimate_infections()
```

My opinion is option 2 is more elegant ("future dates are just another form of missing data and the model doesn't actually need to know when the present is") but option 1 has the advantage more in line with what we've been doing in the package so far.

### PR DESCRIPTION
**Pull request by sbfnk**
## Description

This PR closes #867 and part of #898 whilst enabling addressing #640.

It introduces a `forecast` argument to `estimate_infections()` that is set via `forecast_opts()` (preferred over `add_horizon()` as it allows us to identify the last date in the data - though the `add_horizon()` function exists internally so we could make the switch at a later time if we want).

## Initial submission checklist

- [X] My PR is based on a package issue and I have explicitly linked it.
- [X] I have tested my changes locally (using `devtools::test()` and `devtools::check()`).
- [X] I have added or updated unit tests where necessary.
- [X] I have updated the documentation if required and rebuilt docs if yes (using `devtools::document()`).
- [X] I have followed the established coding standards (and checked using `lintr::lint_package()`).
- [X] I have added a news item linked to this PR.

## After the initial Pull Request 

- [x] I have reviewed Checks for this PR and addressed any issues as far as I am able.

### CODE REVIEW

**Review by jamesmbaazam — CHANGES_REQUESTED** [review]
Thanks for implementing this, Seb.

I made a few comments and suggestions.

File: R/create.R [hunk 1]
```diff
@@ -512,26 +528,33 @@ estimate_early_dynamics <- function(cases, seeding_time) {
...
-  cases <- data[(seeding_time + 1):(.N - horizon)]
-  complete_cases <- create_complete_cases(cases)
-  cases <- cases$confirm
-  accumulate <- data[-(1:seeding_time)]$accumulate
+create_stan_data <- function(data, seeding_time, rt, gp, obs, backcalc,
+                             shifted_cases, forecast) {
+  cases <- data[(seeding_time + 1):.N]
+  cases[, lookup := seq_len(.N)]
+  cases_time <- cases[!is.na(confirm), lookup]
+  imputed_time <- cases[!(accumulate), lookup]
```
[the remark this answers is not in the document]
Comment (sbfnk) [reply R/create.R:536]: I agree but - should we then also rename `cases_time` to `case_times`?

File: R/create.R [hunk 2]
```diff
@@ -543,7 +566,7 @@ create_stan_data <- function(data, seeding_time,
   # calculate prior infections and fit early growth
   stan_data <- c(
     stan_data,
-    estimate_early_dynamics(cases, seeding_time)
+    estimate_early_dynamics(confirmed_cases, seeding_time)
```
[the remark this answers is not in the document]
Comment (sbfnk) [reply R/create.R:569]: Sorry, I shifted things around a bit. Before we had

```r
  cases <- data[(seeding_time + 1):(.N - horizon)]
  cases <- cases$confirm
```

and now we have

```r
  cases <- data[(seeding_time + 1):.N]
  confirmed_cases <- cases[1:(.N - forecast$horizon)]$confirm
```

So I think what was `cases` before is the same as what is `confirmed_cases` now.

File: R/opts.R [hunk 3]
```diff
@@ -1064,7 +1064,38 @@ stan_opts <- function(object = NULL,
   return(opts)
 }
 
-#' Return an _opts List per Region
+#' Forecast options
+#' @description `r lifecycle::badge("stable")`
```
[the remark this answers is not in the document]
Comment (sbfnk) [reply R/opts.R:1068]: Probably not, will change.

File: R/opts.R [hunk 4]
```diff
@@ -1064,7 +1064,38 @@ stan_opts <- function(object = NULL,
...
+#' settings can be supplied which override the defaults.
+#'
+#' @param horizon Numeric, defaults to 7. Number of days into the future to
+#' forecast.
+#' @param accumulate Integer, the number of days to accumulate in forecasts, if
+#'   any. If not given and observations are accumulated at constant frequency in
+#'   the data used for fitting then the same accumulation will be used in
+#'   forecasts unless set explicitly here.
+#' @return A `<forecast_opts>` object of forecast setting.
+#' rstan functions.
```
[the remark this answers is not in the document]
Comment (sbfnk) [reply R/opts.R]: yep

File: tests/testthat/test-gp_opts.R [hunk 5]
```diff
@@ -1,44 +0,0 @@
-test_that("gp_opts returns correct default values", {
```
[the remark this answers is not in the document]
Comment (sbfnk) [reply tests/testthat/test-gp_opts.R:1]: yes, will revert.

**Review by sbfnk — COMMENTED** [review]
Thanks, @jamesmbaazam. Did the new identification of gaps look OK to you (I think you weren't convinced of the idea in the corresponding Issue)? https://github.com/epiforecasts/EpiNow2/blob/8003d06fda877745f7242e09e0355533f1d3bb97/R/create.R#L447

**Comment by jamesmbaazam** [comment]
> I agree but - should we then also rename cases_time to case_times? 

That would be better too.

File: R/create.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (jamesmbaazam) [reply R/create.R:569]: I see now. Thanks.

File: R/create.R (hunk 1 above)
[the remark this answers is not in the document]
Comment (sbfnk) [reply R/create.R]: https://github.com/epiforecasts/EpiNow2/pull/901/commits/e6cdb09ece341736258cad45608efed14af58ce8 and  https://github.com/epiforecasts/EpiNow2/pull/901/commits/8302c5cf7e26712f4850c2b65f88d0933c1f33c8

File: tests/testthat/test-gp_opts.R (hunk 5 above)
[the remark this answers is not in the document]
Comment (sbfnk) [reply tests/testthat/test-gp_opts.R]: [`d9c1430` (#901)](https://github.com/epiforecasts/EpiNow2/pull/901/commits/d9c14302f51fa75f98c03d3ed7f5ee0d11c98852)

File: R/opts.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (jamesmbaazam) [reply R/opts.R:1072]: Okay, that clarifies it. Thanks.

**Review by jamesmbaazam — APPROVED** [review]

**Pull request merged**